openfl/experimental/workflow/component/envoy/envoy.py

```diff
@@ -0,0 +1,224 @@
+# Copyright 2020-2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+"""Experimental Envoy module."""
+
+import logging
+import sys
+import time
+import uuid
+from concurrent.futures import ThreadPoolExecutor
+from pathlib import Path
+from typing import Optional, Union
+
+from openfl.experimental.workflow.federated import Plan
+from openfl.experimental.workflow.transport.grpc.director_client import DirectorClient
+from openfl.experimental.workflow.transport.grpc.exceptions import EnvoyNotFoundError
+from openfl.utilities.workspace import ExperimentWorkspace
+
+logger = logging.getLogger(__name__)
+
+
+class Envoy:
+    """Envoy class. The Envoy is a long-lived entity that runs on collaborator
+    nodes connected to the Director.
+
+    Attributes:
+        name (str): The name of the envoy.
+        envoy_config (Optional[Path]): Path to envoy_config.yaml
+        tls (bool, optional): A flag indicating if TLS should be used for
+                connections. Defaults to True.
+        root_certificate (Optional[Union[Path, str]]): The path to the root certificate
+            for TLS.
+        private_key (Optional[Union[Path, str]]): The path to the private key for TLS.
+        certificate (Optional[Union[Path, str]]): The path to the certificate for TLS.
+        director_client (DirectorClient): The director client.
+        install_requirements (bool): A flag indicating if the requirements
+            should be installed.
+        is_experiment_running (bool): A flag indicating if an experiment is
+            running.
+        executor (ThreadPoolExecutor): The executor for running tasks.
+        plan(str): Path to plan.yaml
+        _health_check_future (object): The future object for the health check.
+    """
+
+    DEFAULT_RETRY_TIMEOUT_IN_SECONDS = 5
+
+    def __init__(
+        self,
+        *,
+        envoy_name: str,
+        director_host: str,
+        director_port: int,
+        envoy_config: Optional[Path] = None,
+        root_certificate: Optional[Union[Path, str]] = None,
+        private_key: Optional[Union[Path, str]] = None,
+        certificate: Optional[Union[Path, str]] = None,
+        tls: bool = True,
+        install_requirements: bool = True,
+    ) -> None:
+        """Initialize a envoy object.
+
+        Args:
+            envoy_name (str): The name of the envoy.
+            director_host (str): The host of the director.
+            director_port (int): The port of the director.
+            envoy_config (Optional[Path]): Path to envoy_config.yaml
+            root_certificate (Optional[Union[Path, str]]): The path
+                to the root certificate for TLS. Defaults to None.
+            private_key (Optional[Union[Path, str]]): The path to
```

Comment:
Same as for the Director, the `private_key` should be required if TLS is enabled

Reply:
Same as https://github.com/securefederatedai/openfl/pull/1190#discussion_r1875951703